include/envoy/upstream/cluster_manager.h

```diff
@@ -309,6 +345,24 @@ class ClusterManager {
   virtual const ClusterRequestResponseSizeStatNames&
   clusterRequestResponseSizeStatNames() const PURE;
   virtual const ClusterTimeoutBudgetStatNames& clusterTimeoutBudgetStatNames() const PURE;
+
+  /**
+   * Request an on-demand discovery of a cluster with a passed name. Passed ODCDS may be used to
+   * perform the discovery process in the main thread if there is no discovery going on for this
+   * cluster. The passed callback will be invoked when the cluster is added and warmed up. It is
+   * expected that the callback will be destroyed when it is invoked. To cancel the discovery,
+   * destroy the returned handle and the callback.
+   *
+   * This function is thread-safe.
+   *
+   * @param odcds is a pointer to ODCDS used for discovery. Must not be a nullptr.
+   * @param name is the name of the cluster to be discovered.
+   * @param callback will be called when the discovery is finished.
+   * @return ClusterDiscoveryCallbackHandlePtr the discovery process handle.
+   */
+  virtual ClusterDiscoveryCallbackHandlePtr
+  requestOnDemandClusterDiscovery(OdCdsApiSharedPtr odcds, const std::string& name,
```

Comment:
Starting from a very high level: how does one allocate an OdCdsAPI, and why does one allocate it? It seems like this should all be handled by the cluster manager? Is the issue config? Should there be a CM API to allocate an on demand handle or something like that?

Reply:
> Starting from a very high level: how does one allocate an OdCdsAPI, and why does one allocate it?

An extension could have a config source in its configuration, so it could be used to create an OdCdsApi. Extension config creation happens in main thread, so this looks like a good moment to create an instance. The source config could be different from cds_config in bootstrap, which may be the reason why a separate OdCdsApi needs to be allocated.

> It seems like this should all be handled by the cluster manager? Is the issue config? Should there be a CM API to allocate an on demand handle or something like that?

Could be I suppose. I can see adding something like:

```c++
class OdCdsApiHandle {
public:
  virtual ~OdCdsApiHandle() = default;
};

using OdCdsApiHandleSharedPtr = std::shared_ptr<OdCdsApiHandle>;

…

virtual OdCdsApiHandleSharedPtr ClusterManager::allocateOdCdsApi(…);
virtual ClusterDiscoveryCallbackHandlePtr
  requestOnDemandClusterDiscovery(OdCdsApiHandleSharedPtr odcds, …);
```

So I'm not sure. It's replacing an interface with one function for another interface with no functions + one more function in CM.

Reply:
Yeah ^ is what I would do. It's cleaner because now everything is contained within the CM and the handle is effectively opaque and can be de-duped based on config.

/wait

Reply:
I'm halfway through the refactor, so I have couple of notes:

- I have kept the `OdCdsApiImpl` class for now, it just does not subclass the already gone `OdCdsApi` interface. I think I'll rename the class to `OdCdsApi` later, since it's not an impl of anything anymore.
- Having an opaque `OdCdsApiHandle` classes basically forced me to write a function like follows inside `cluster_manager_impl.cc`:

```c++
OdCdsApiImpl& getOdCdsApiFromHandle(OdCdsApiHandleSharedPtr const& handle) {
  auto impl = std::dynamic_pointer_cast<OdCdsApiHandleImpl> (handle);
  ASSERT(impl != nullptr);
  return impl->get();
}
```

which feels kind of ew, but I have seen `std::dynamic_pointer_cast` being used in Envoy, so I guess this is ok.

This also means that it probably became difficult to keep the unit tests I wrote (the `ODCDTest` class in `cluster_manager_impl_test.cc`), because I can't mock OdCds. Passing a mock of the handle will trigger an assertion. I'll comment them out for now.

Reply:
You shouldn't need any casts. Just access the functionality from the handle (i.e. move the dynamic fetch functionality to the handle).

Reply:
Oh. My understanding was that you wanted the handle class to be completely opaque, so no other virtual functions other than the destructor. Adding a dynamic fetch functionality to the handle means renaming the old `OdCdsApi` interface to `OdCdsApiHandle`.

Reply:
Basically what I'm after is for the entire functionality to be exposed by the CM in isolation. So I think API handle -> fetch -> fetch handle makes sense. The API handle should be obtained from the CM itself based on config (which should be de-duped).

Reply:
The only wrinkle I can see here is that you can receive an API handle from CM and then instead of doing `cm.requestOnDemandClusterDiscovery(api_handle, …)`, you can also do `api_handle.updateOnDemand(clusterName)`. So maybe expose the discovery functionality only through the handle and drop the `requestOnDemandClusterDiscovery` function from CM interface? Or it would more like - move `requestOnDemandClusterDiscovery` from CM to API handle.

Reply:
Yes exactly, the only CM API is to get an API handle, then the request is done through the API handle. This makes sure the entire abstraction is hidden inside the CM, and then it's just an implementation detail of code sharing between CDS and ODCDS.